cloud/amazon/ec2_vpc_dhcp_options.py

```diff
@@ -128,105 +157,204 @@
-        module.exit_json(changed=False)
-  
-    if new_options['netbios-node-type']:
-        new_options['netbios-node-type'] = new_options['netbios-node-type'][0]
-  
-    if new_options['domain-name']:
-        new_options['domain-name'] = new_options['domain-name'][0]
-  
-    if not module.check_mode:
-        dhcp_option = connection.create_dhcp_options(new_options['domain-name'],
-                new_options['domain-name-servers'], new_options['ntp-servers'],
-                new_options['netbios-name-servers'], new_options['netbios-node-type'])
+
+    # First check if we were given a dhcp_options_id
+    if not params['dhcp_options_id']:
+    # No, so create new_options from the parameters
+        if params['dns_servers'] != None:
+            new_options['domain-name-servers'] = params['dns_servers']
+        if params['netbios_name_servers'] != None:
+            new_options['netbios-name-servers'] = params['netbios_name_servers']
+        if params['ntp_servers'] != None:
+            new_options['ntp-servers'] = params['ntp_servers']
+        if params['domain_name'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['domain-name'] = [ params['domain_name'] ]
+        if params['netbios_node_type'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['netbios-node-type'] = [ str(params['netbios_node_type']) ]
+    # If we were given a vpc_id then we need to look at the options on that
+    if params['vpc_id']:
+        existing_options = fetch_dhcp_options_for_vpc(connection, params['vpc_id'])
+        # if we've been asked to inherit existing options, do that now
+        if params['inherit_existing']:
+            for option in [ 'domain-name-servers', 'netbios-name-servers', 'ntp-servers', 'domain-name', 'netbios-node-type']:
+                if existing_options.options.get(option) and (not new_options[option] or [''] == new_options[option]):
+                    new_options[option] = existing_options.options.get(option)
+
+        # Do the vpc's dhcp options already match what we're asked for? if so we are done
+        if existing_options and new_options == existing_options.options:
+            found = True
+            dhcp_option = existing_options
+            module.exit_json(changed=changed, new_options=new_options, dhcp_options_id=existing_options.id)
+
+        # If no vpc_id was given, or the options don't match then look for an existing set using tags
+        found, dhcp_option = match_dhcp_options(connection, params['tags'], new_options)
+
+    # Now let's cover the case where there are existing options that we were told about by id
+    # If a dhcp_options_id was supplied we don't look at options inside, just set tags (if given)
+    else:
+        supplied_options = connection.get_all_dhcp_options(dhcp_options_ids=params['dhcp_options_id'])
```

Comment:
Thanks for adding the additional check, but I wasn't clear in my original response, sorry. If you pass in a `dhcp_options_ids` parameter to `get_all_dhcp_options` then it will throw an exception:

```
$ python
>>> import boto.vpc
>>> c = boto.vpc.connect_to_region('us-east-1')
>>> c.get_all_dhcp_options(dhcp_options_ids='dopt-12345678')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/boto/vpc/__init__.py", line 1228, in get_all_dhcp_options
    [('item', DhcpOptions)])
  File "/usr/local/lib/python2.7/dist-packages/boto/connection.py", line 1186, in get_list
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidDhcpOptionID.NotFound</Code><Message>The dhcpOption ID 'dopt-12345678' does not exist</Message></Error></Errors><RequestID>5b749029-e7d2-449e-9960-f0a928eab648</RequestID></Response>
```

I think changing the code to this would solve the problems:

`supplied_options = [d for d in connection.get_all_dhcp_options() if d.id == params['dhcp_options_id']`

(I've found myself writing this type of boilerplate a lot, unfortunately...)